spark3/src/main/java/org/apache/iceberg/spark/procedures/RewriteManifestsProcedure.java

```diff
@@ -0,0 +1,108 @@
+
+import org.apache.iceberg.actions.Actions;
+import org.apache.iceberg.actions.RewriteManifestsAction;
+import org.apache.iceberg.actions.RewriteManifestsActionResult;
+import org.apache.iceberg.spark.procedures.SparkProcedures.ProcedureBuilder;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.iceberg.catalog.ProcedureParameter;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+class RewriteManifestsProcedure extends BaseProcedure {
+
+  private static final ProcedureParameter[] PARAMETERS = new ProcedureParameter[]{
+      ProcedureParameter.required("namespace", DataTypes.StringType),
+      ProcedureParameter.required("table", DataTypes.StringType),
+      ProcedureParameter.optional("spec_id", DataTypes.IntegerType),
+      ProcedureParameter.optional("use_caching", DataTypes.BooleanType)
+  };
+
+  // counts are not nullable since the action result is never null
+  private static final StructType OUTPUT_TYPE = new StructType(new StructField[]{
+      new StructField("num_rewritten_manifests", DataTypes.IntegerType, false, Metadata.empty()),
+      new StructField("num_added_manifests", DataTypes.IntegerType, false, Metadata.empty())
+  });
+
+  public static ProcedureBuilder builder() {
+    return new BaseProcedure.Builder<RewriteManifestsProcedure>() {
+      @Override
+      protected RewriteManifestsProcedure doBuild() {
+        return new RewriteManifestsProcedure(tableCatalog());
+      }
+    };
+  }
+
+  private RewriteManifestsProcedure(TableCatalog tableCatalog) {
+    super(tableCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    String namespace = args.getString(0);
+    String tableName = args.getString(1);
+    Integer specId = args.isNullAt(2) ? null : args.getInt(2);
+    Boolean useCaching = args.isNullAt(3) ? null : args.getBoolean(3);
+
+    return withIcebergTable(namespace, tableName, table -> {
+      Actions actions = Actions.forTable(table);
+
+      RewriteManifestsAction action = actions.rewriteManifests();
+
+      if (specId != null) {
+        action.specId(specId);
+      }
+
+      if (useCaching != null) {
+        action.useCaching(useCaching);
+      }
+
+      RewriteManifestsActionResult result = action.execute();
+
+      Object[] outputValues = new Object[OUTPUT_TYPE.size()];
+      outputValues[0] = result.deletedManifests().size();
+      outputValues[1] = result.addedManifests().size();
+      GenericInternalRow outputRow = new GenericInternalRow(outputValues);
```

Comment:
Not sure if I already mentioned this, but we might want to add a helper method to create a generic row.

Reply:
Done in a separate PR.